### PR TITLE
CHE-55: Adding io.fabric8 dependencies

### PR DESCRIPTION
**Pull request by ibuziuk**
The change is coupled with PR - https://github.com/eclipse/che/pull/3381
While building "plugin-docker" without "skip-enforce" profile I got the following error:

```
[WARNING] Used undeclared dependencies found:
[WARNING]    io.fabric8:kubernetes-model:jar:1.0.63:compile
[WARNING]    io.fabric8:kubernetes-client:jar:1.4.26:compile
[INFO] Add the following to your pom to correct the missing dependencies: 
[INFO] 
<dependency>
  <groupId>io.fabric8</groupId>
  <artifactId>kubernetes-model</artifactId>
  <version>1.0.63</version>
</dependency>
<dependency>
  <groupId>io.fabric8</groupId>
  <artifactId>kubernetes-client</artifactId>
  <version>1.4.26</version>
</dependency> 
```
In order to make it work I had to add those dependencies to pom.xml even though only "openshift-client" is required. Both kubernetes-client & openshift-client are located in the same repository and the CQ allows to proceed with checkin. However, kubernetes-model is located in different one and no CQ has been created for it. Do we need to create one even though it is not used directly ?  

mvn dependency:tree
```
[INFO] +- io.fabric8:openshift-client:jar:1.4.26:compile
[INFO] |  \- io.fabric8:kubernetes-client:jar:1.4.26:compile
[INFO] |     +- io.fabric8:kubernetes-model:jar:1.0.63:compile
[INFO] |     |  \- com.fasterxml.jackson.module:jackson-module-jaxb-annotations:jar:2.7.5:compile
[INFO] |     +- com.squareup.okhttp3:okhttp:jar:3.4.1:compile
[INFO] |     |  \- com.squareup.okio:okio:jar:1.9.0:compile
[INFO] |     +- com.squareup.okhttp3:logging-interceptor:jar:3.4.1:compile
[INFO] |     +- com.squareup.okhttp3:okhttp-ws:jar:3.4.1:compile
[INFO] |     +- org.slf4j:jul-to-slf4j:jar:1.7.6:compile
[INFO] |     +- com.fasterxml.jackson.dataformat:jackson-dataformat-yaml:jar:2.5.0:compile
[INFO] |     +- com.fasterxml.jackson.core:jackson-databind:jar:2.5.0:compile
[INFO] |     |  \- com.fasterxml.jackson.core:jackson-annotations:jar:2.5.0:compile
[INFO] |     +- com.fasterxml.jackson.core:jackson-core:jar:2.5.0:compile
[INFO] |     +- io.fabric8:zjsonpatch:jar:0.2.3:compile
[INFO] |     \- com.github.mifmif:generex:jar:1.0.1:compile
[INFO] |        \- dk.brics.automaton:automaton:jar:1.11-8:compile
```

CQ - https://dev.eclipse.org/ipzilla/show_bug.cgi?id=12209
Jira - https://issues.jboss.org/browse/CHE-55

### CODE REVIEW

**Comment by skabashnyuk** [comment]
1. Are you sure you've made cq for all three declared decencies?
2. Are you sure that all transitive dependencies are already approved?

**Comment by ibuziuk** [comment]
@skabashnyuk For kubernetes-client & openshift-client the CQ is opened and reviewer wrote "Please proceed with checkin". For "kubernetes-model" there is no CQ opened and we do not use this library directly (kubernetes-client just depends on it). However, without specifying it in the pom.xm explicitly the build fails with: 

```
[WARNING] Used undeclared dependencies found:
[WARNING]    io.fabric8:kubernetes-model:jar:1.0.63:compile
[INFO] Add the following to your pom to correct the missing dependencies: 
[INFO] 
<dependency>
  <groupId>io.fabric8</groupId>
  <artifactId>kubernetes-model</artifactId>
  <version>1.0.63</version>
</dependency>
```
Should we open a CQ for it even though it is not used directly ?

**Comment by skabashnyuk** [comment]
if it's not needed just exclude it like this https://github.com/eclipse/che-dependencies/blob/master/pom.xml#L195-L204
what about transitive dependencies?

**Comment by ibuziuk** [comment]
@skabashnyuk Do we really need to open CQ for each transitive dependency?  I have done some CQs in the past and there were no requests for transitive dependencies processing. For this change we have only one CQ - https://dev.eclipse.org/ipzilla/show_bug.cgi?id=12209

**Comment by skabashnyuk** [comment]
yes- if you adding new library and this library added new transitive dependencies.
you have a couple of ways to check it. 
1. Use dependency tree and goes over the all list of transitive dependencies and compare to the list of already approved CQs. Sometimes this method can skip some libraries. 
2. Most productive way is to compare API war before and after adding new dependencies. All new jars have to be approved with CQ

**Comment by ibuziuk** [comment]
@skabashnyuk could you please take a look at -https://wiki.eclipse.org/Development_Resources/Contribution_Questionnaire#Third_Party_Libraries

As far as I am concerned, we do not need to open any CQs for transitive dependencies as long as they are  not used directly. So, the only CQ that should be opened is for **kubernetes-model** - https://dev.eclipse.org/ipzilla/show_bug.cgi?id=12440

**Comment by skabashnyuk** [comment]
Can you exclude all transitive dependencies that has not Che CQ?  
We do need to create CQ if transitive dependency comes with assembly. Even if it not used.

**Comment by ibuziuk** [comment]
@skabashnyuk  probably we can exclude some of the dependencies
@l0rd the following dependencies can not be excluded for sure, so we need to open CQs for them:

1. com.squareup.okhttp3:okhttp:jar:3.4.1
2. com.squareup.okhttp3:logging-interceptor:jar:3.4.1
3. com.squareup.okio:okio:jar:1.9.0

I have excluded the following dependencies and was able to build "openshift-connector" branch and run the test:
- com.fasterxml.jackson.module:jackson-module-jaxb-annotations:jar:2.7.5
- com.squareup.okhttp3:okhttp-ws:jar:3.4.1
- io.fabric8:zjsonpatch
- com.github.mifmif:generex

Need to do more testing in order to figure whether those libs are required in the runtime

**Comment by l0rd** [comment]
I've created the following CQs:

- kubernetes-model Version: 1.0.63 => https://dev.eclipse.org/ipzilla/show_bug.cgi?id=12440
- jackson-module-jaxb-annotations Version: 2.7.5 => https://dev.eclipse.org/ipzilla/show_bug.cgi?id=12441
- okhttp Version: 3.4.1 => https://dev.eclipse.org/ipzilla/show_bug.cgi?id=12442
- okio Version: 1.9.0 => https://dev.eclipse.org/ipzilla/show_bug.cgi?id=12443

Maven artifacts `com.squareup.okhttp3:okhttp:jar:3.4.1`, `com.squareup.okhttp3:logging-interceptor:jar:3.4.1` and `com.squareup.okhttp3:okhttp-ws:jar:3.4.1` are all submodules of okttp version 3.4.1 (source code is on the same github repo for the 3 artifacts). Therefore they should not need separates CQs.

**Comment by skabashnyuk** [comment]
@l0rd  ok let us know please when they will be approved

**Comment by ibuziuk** [comment]
@l0rd we need to open yet another CQ (got a bunch of NoClassDefFound error while testing):
- io.fabric8:zjsonpatch:jar:0.2.3

**Comment by ibuziuk** [comment]
@l0rd it looks like the only dependency that can be excluded is  **com.github.mifmif:generex:jar:1.0.1** and it's transient **dk.brics.automaton:automaton:jar:1.11-8**. At least I had no issues during testing. PR was updated with those lib exclusion. For all other dependencies except **io.fabric8:zjsonpatch:jar:0.2.3** CQs have already been opened

**Comment by l0rd** [comment]
I've opened a new CQ for zjsonpatch: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=12451

Here is the list of current CQs. I will check a CQ as soon as with have an approval:
- [x] kubernetes-model Version: 1.0.65 [CQ](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=12440)
- [x] jackson-module-jaxb-annotations Version: 2.7.5 [CQ](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=12441)
- [x] okhttp Version: 3.4.1 [CQ](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=12442)
- [x] okio Version: 1.9.0 [CQ](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=12443)
- [x] zjsonpatch Version: 0.3.0 [CQ](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=12451)
- [x] kubernetes-client Version: 1.4.31 [CQ](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=12506)

**Comment by ibuziuk** [comment]
PR has been updated due to zjsonpatch 0.2.3 copyright header issue:
```
[INFO] +- io.fabric8:kubernetes-client:jar:1.4.31:compile
[INFO] |  +- com.squareup.okhttp3:okhttp:jar:3.4.1:compile
[INFO] |  |  \- com.squareup.okio:okio:jar:1.9.0:compile
[INFO] |  +- com.squareup.okhttp3:logging-interceptor:jar:3.4.1:compile
[INFO] |  +- com.squareup.okhttp3:okhttp-ws:jar:3.4.1:compile
[INFO] |  +- org.slf4j:jul-to-slf4j:jar:1.7.6:compile
[INFO] |  +- com.fasterxml.jackson.dataformat:jackson-dataformat-yaml:jar:2.5.0:compile
[INFO] |  +- com.fasterxml.jackson.core:jackson-databind:jar:2.5.0:compile
[INFO] |  |  \- com.fasterxml.jackson.core:jackson-annotations:jar:2.5.0:compile
[INFO] |  +- com.fasterxml.jackson.core:jackson-core:jar:2.5.0:compile
[INFO] |  +- io.fabric8:zjsonpatch:jar:0.3.0:compile
[INFO] |  \- com.github.mifmif:generex:jar:1.0.1:compile
[INFO] |     \- dk.brics.automaton:automaton:jar:1.11-8:compile
[INFO] +- io.fabric8:kubernetes-model:jar:1.0.65:compile
```

CQs for "kubernetes-client", "kubernetes-model" & "zjsonpatch" has been updated

**Comment by skabashnyuk** [comment]
are you going to exclude 

```
[INFO] |  \- com.github.mifmif:generex:jar:1.0.1:compile
[INFO] |     \- dk.brics.automaton:automaton:jar:1.11-8:compile
```
?

**Comment by ibuziuk** [comment]
@skabashnyuk yup, it is excluded

**Comment by skabashnyuk** [comment]
Any news?

**Comment by ibuziuk** [comment]
@skabashnyuk No news :-( Waiting for **zjsonpatch** & **kubernetes-model** CQ clarifications from eclipse IP team

**Comment by ibuziuk** [comment]
@skabashnyuk @l0rd Good news! CQ for **zjsonpatch** has been approved and we got an allowance to proceed with  **kubernetes-model-1.0.65** checkin. So, I guess the PR can be safely merged now: 

```
[INFO] +- io.fabric8:kubernetes-client:jar:1.4.31:compile
[INFO] |  +- com.squareup.okhttp3:okhttp:jar:3.4.1:compile
[INFO] |  |  \- com.squareup.okio:okio:jar:1.9.0:compile
[INFO] |  +- com.squareup.okhttp3:logging-interceptor:jar:3.4.1:compile
[INFO] |  +- com.squareup.okhttp3:okhttp-ws:jar:3.4.1:compile
[INFO] |  +- org.slf4j:jul-to-slf4j:jar:1.7.6:compile
[INFO] |  +- com.fasterxml.jackson.dataformat:jackson-dataformat-yaml:jar:2.5.0:compile
[INFO] |  +- com.fasterxml.jackson.core:jackson-databind:jar:2.5.0:compile
[INFO] |  |  \- com.fasterxml.jackson.core:jackson-annotations:jar:2.5.0:compile
[INFO] |  +- com.fasterxml.jackson.core:jackson-core:jar:2.5.0:compile
[INFO] |  \- io.fabric8:zjsonpatch:jar:0.3.0:compile
[INFO] +- io.fabric8:kubernetes-model:jar:1.0.65:compile
[INFO] |  \- javax.validation:validation-api:jar:1.1.0.Final:compile
[INFO] +- io.fabric8:openshift-client:jar:1.4.31:compile
```

**Review by l0rd — APPROVED** [review]

**Review by benoitf — APPROVED** [review]
Need to wait tomorrow before merging. (release will occur today)

**Comment by ibuziuk** [comment]
PR has been rebased

**Review by skabashnyuk — APPROVED** [review]

**Comment by skabashnyuk** [comment]
We good to go.

**Pull request merged**
